### PR TITLE
Add `_.replaceAll`

### PR DESCRIPTION
**Pull request by mcharytoniuk**
This pull request contains `_.replaceAll` static function. Default behavior of JavaScript's `String.prototype.replace` is to replace only first occurence of a string (unless `RegExp` is used) and `_.replaceAll` replaces all occurences. 

```JavaScript
_.replaceAll('b-c-d', '-', '.');
// => 'b.c.d'
```

It could be eventually further improved in the future to work in a chain like this:

```JavaScript
_(['a.b.c', 'def']).replaceAll('.', '..').value(); 
// => ['a..b..c', 'def']
```

### CODE REVIEW

**Comment by StreetStrider** [comment]
@mcharytoniuk, how to use RegExp `i` flag in this function?

**Comment by mcharytoniuk** [comment]
The goal of this function is not to use RegExp to replace strings. `RegExp ` it's an implementation detail here.

**Comment by jdalton** [comment]
Related to #1084.

**Comment by jdalton** [comment]
Hi @mcharytoniuk. Do you have a use case in mind?

**Comment by mcharytoniuk** [comment]
I just use this pretty often and usually end up with something like this:
https://github.com/goreutils/gore-gulp/blob/69fa3d57b7a35beb8936aaec5ec559b2f64570b9/gore-gulp/webpack.js#L68

I find this kludgy, unreliable and repetetive. I would love to see such function in a popular library so it would be much more tested and reliable. Possibly some other parts of lodash could benefit from it in the future and stay more maintainable. It allows to do this simple operation in one function call instead of creating loops or regular expressions. I use lodash in most of my projects so I would not have to repeat myself.

**Comment by jdalton** [comment]
@mcharytoniuk can you create a [jsperf](https://jsperf.com) of that and this implementation.

**Comment by mcharytoniuk** [comment]
Yes, I can but loop + indexOf implementation is not generic because it would fail in the following case:

```JavaScript
replace('a/b/c', '/', '//');
```

It would end in an endless loop producing something like this:

```JavaScript
'a/////////////////////////////(...)'
```

I would have to store the index of last replacement somewhere and look for occurence after given replacement and do some other string manipulations which would be compilcated for such thing. I have also this implementation in mind:

```JavaScript
function replaceAll(string, pattern, replacement) {
  string = baseToString(string);
  pattern = baseToString(pattern);

  if (!pattern) {
    return string;
  }

  replacement = baseToString(replacement);
  
  string = string.split(pattern);

  return string.join(replacement);
}
```

I'll creat jsperf soon.

**Comment by mcharytoniuk** [comment]
jsperf: http://jsperf.com/lodash-string-replace-all

**Comment by jdalton** [comment]
@mcharytoniuk Thanks! This won't make it into 3.7.0 but may make it into the version after. I'll kick it around a bit more.

**Comment by mcharytoniuk** [comment]
No problem. :)

File: lodash.src.js [hunk 1]
```diff
@@ -10610,6 +10610,40 @@
...
+     * // => 'a//b//c'
+     *
+     * _.replaceAll('aabbaabbaacc', 'aa', 'd');
+     * // => 'dbbdbbdcc'
+     */
+    function replaceAll(string, pattern, replacement) {
+      string = baseToString(string);
+      pattern = baseToString(pattern);
+
+      if (!pattern) {
```
Comment (megawac) [inline lodash.src.js]: this doesn't seem strictly necessary, replace all(str, '', seperator) seems valid

Comment (jdalton) [inline lodash.src.js]: Yap, good point.

File: test/test.js [hunk 2]
```diff
@@ -13064,6 +13064,29 @@
...
+      strictEqual(_.replaceAll('a/b/c', '/', '//'), 'a//b//c');
+    });
+
+    test('should coerce `pattern` and `replacement` to a string', 1, function() {
+      strictEqual(_.replaceAll('123456', 3, '4'), '124456');
+    });
+
+    test('should coerce `replacement` to a string', 1, function() {
+      strictEqual(_.replaceAll('123', '2', { 'toString': _.constant('4') }), '143');
+    });
```
Comment (megawac) [inline test/test.js]: could you add a test for a non global regexp? afterwards I'm +1

Comment (megawac) [inline test/test.js]: also I suppose that case should be discussed

Comment (jdalton) [inline test/test.js]: The param specifies it as a string, I'd assume if someone wanted regexes they'd use `String#replace`.

Comment (megawac) [inline test/test.js]: I was mentioning it for chain cases

Comment (jdalton) [inline test/test.js]: Still not sure I get what regexes have to do with it.

File: lodash.src.js (hunk 1 above)
Comment (jdalton) [inline lodash.src.js]: ping @mcharytoniuk.

Comment (mcharytoniuk) [inline lodash.src.js]: Right, I'll make a commit in a second.

Comment (mcharytoniuk) [inline lodash.src.js]: Sorry for a delay.

Pure JavaScript gives output like that:
```JavaScript
"test".replace("", "/");
// => "/test"
```

So to be consistent `replaceAll` should work like this, which is strange IMO:
```JavaScript
_.replaceAll("test", "", "/");
// => "/t/e/s/t"
```

**Comment by jdalton** [comment]
Thanks for this @mcharytoniuk! I've been keeping this PR around for a bit to see if I'd dig it anymore but haven't really gotten excited about it. I'm going to close it until demand gets a little stronger and then cherry-pick this in (so you get credit) at that time.

**Comment by sbryfcz** [comment]
Maybe bring this back since more and more string functions are getting added? Would love this in lodash. Seems like I'm always having to add this function to my project's utility libraries.

**Comment by catalint** [comment]
+1

**Comment by ellis** [comment]
I would love to see this in lodash too!  I regularly need a simple replaceAll() function like this.

**Comment by river-jade** [comment]
I'd love to see this. replaceAll seems to be far more useful in my experience than replace (first). The built-in version works for most cases, but it's clunky imo.

**Comment by nevir** [comment]
This would be particularly helpful.  It's something that comes up relatively often, and is likely to reduce errors (such as people forgetting to `_.escapeRegExp`)

**Comment by skfd** [comment]
That would be a tasty feature. People periodically come to [ask for it](https://github.com/lodash/lodash/issues?utf8=%E2%9C%93&q=replaceall) now and then.

**Pull request closed without merging**